### PR TITLE
Enable nplb on Linux.

### PR DESCRIPTION
**Pull request by y4vor**
b/378879686

### CODE REVIEW

File: cobalt/build/configs/BUILD.gn [hunk 1]
```diff
@@ -23,10 +23,7 @@ config("cobalt_config") {
     configs += [ "//${starboard_path}/platform_configuration" ]
   }
 
-  defines = [
-    "ENABLE_BUILDFLAG_IS_COBALT",
-    "OS_STARBOARD",
-  ]
+  defines = [ "ENABLE_BUILDFLAG_IS_COBALT" ]
```
Comment (andrewsavage1) [inline cobalt/build/configs/BUILD.gn:26]: I know it's unrelated, but would you mind removing this one as well? It's also defined conditionally below

File: starboard/common/test_main.cc [hunk 2]
```diff
@@ -27,12 +26,19 @@ int InitAndRunAllTests(int argc, char** argv) {
 }
 }  // namespace
 
-// When we are building Evergreen we need to export SbEventHandle so that the
-// ELF loader can find and invoke it.
+#if defined(OS_STARBOARD)
```
Comment (andrewsavage1) [inline starboard/common/test_main.cc:29]: Let's use `BUILDFLAG(IS_STARBOARD)` unless there's an explicit reason for this file to not use a //build header

File: cobalt/build/configs/BUILD.gn (hunk 1 above)
Comment (y4vor) [reply cobalt/build/configs/BUILD.gn:26]: Done

File: starboard/common/test_main.cc (hunk 2 above)
Comment (y4vor) [reply starboard/common/test_main.cc:29]: The BUILDFLAG doesn't work for starboard/common as it doesn't depend on //base.

Building nplb gives the following error:
../../starboard/common/test_main.cc:29:5: error: function-like macro 'BUILDFLAG' is not defined
#if BUILDFLAG(IS_STARBOARD) 

The options I was considering were defined(STARBOARD) vs defined(OS_STARBOARD).
The third_party, that doesn't depend on //base has some variations of the OS_FOO. I see OS_ANDROID, ABSL_OS_ANDROID, ARMV8_OS_ANDROID.

I'm thinking defined(OS_STARBOARD) is decent choice here.

Comment (andrewsavage1) [reply starboard/common/test_main.cc:29]: BUILDFLAG is defined in //build (`#include "//build/build_config.h"`), not //base, so it should be fine to include in starboard

File: .github/config/linux.json [hunk 3]
```diff
@@ -14,7 +14,8 @@
         "gin_unittests",
         "media_unittests",
         "content_shell",
-        "cobalt:gn_all"
+        "cobalt:gn_all",
+        "nplb"
     ],
```
Comment (yell0wd0g) [inline .github/config/linux.json:19]: Is it me or this list is not in alphabetical order ? :-)

File: starboard/common/test_main.cc [hunk 4]
```diff
@@ -27,12 +26,19 @@ int InitAndRunAllTests(int argc, char** argv) {
...
 SB_EXPORT STARBOARD_WRAP_SIMPLE_MAIN(InitAndRunAllTests);
+
+#if !SB_IS(EVERGREEN)
+// Define main for non-Evergreen Starboard OS.
+int main(int argc, char** argv) {
+  SbRunStarboardMain(argc, argv, SbEventHandle);
+}
+#endif  // !SB_IS(EVERGREEN)
 #else
+// If the OS is not Starbaord use the regular main
```
Comment (yell0wd0g) [inline starboard/common/test_main.cc:40]: This path would be taken if `OS_STARBOARD` is not defined - but would this file ever be built, considering this folder is  called `//starboard` ? IOW do we need this #else ? (If we're not sure I'd propose leaving a `#error This combination should not happen` or some such.

**Review by yell0wd0g — COMMENTED** [review]
Some minor comments, otherwise this is a step forward.

File: starboard/common/test_main.cc (hunk 4 above)
Comment (y4vor) [reply starboard/common/test_main.cc:40]: Here is where the water gets muddy. This case is actually used by ATV and could be used by other platforms in the future if we use the native OS as the platform but still use Starboard as a library e.g. SbPlayer.

Added a comment to explain that ATV uses it.

File: starboard/common/test_main.cc [hunk 5]
```diff
@@ -27,12 +26,19 @@ int InitAndRunAllTests(int argc, char** argv) {
...
+#if defined(OS_STARBOARD)
+// For the Starboard OS define SbEventHandle as the entry point
 SB_EXPORT STARBOARD_WRAP_SIMPLE_MAIN(InitAndRunAllTests);
+
+#if !SB_IS(EVERGREEN)
+// Define main for non-Evergreen Starboard OS.
+int main(int argc, char** argv) {
+  SbRunStarboardMain(argc, argv, SbEventHandle);
+}
+#endif  // !SB_IS(EVERGREEN)
```
[the remark this answers is not in the document]
Comment (y4vor) [reply starboard/common/test_main.cc:38]: I think  we have both OS_STARBOARD and !OS_STARBOARD.
And under OS_STARBOARD we have main() for the current Linux approach and we don't have main() for Evergreen but both use SbEventHandle.

File: .github/config/linux.json (hunk 3 above)
Comment (y4vor) [reply .github/config/linux.json:19]: Sorted

File: .github/config/linux.json [hunk 6]
```diff
@@ -5,16 +5,17 @@
       "linux-x64x11"
     ],
     "targets": [
```
Comment (niranjanyardi) [inline .github/config/linux.json:7]: there's no need to modify .github/config/linux.json for adding nplb

cobalt/BUILD.gn already has nplb as one of it's targets and we are already building that for linux
https://source.corp.google.com/h/github/youtube/cobalt/+/main:cobalt/BUILD.gn;l=21

File: starboard/common/test_main.cc (hunk 2 above)
Comment (andrewsavage1) [reply starboard/common/test_main.cc:29]: Marking as unresolved

Comment (y4vor) [reply starboard/common/test_main.cc:29]: Done

File: .github/config/linux.json (hunk 6 above)
Comment (y4vor) [reply .github/config/linux.json:7]: oh, interesting. 
Where do we specify the tests to be run?

Comment (andrewsavage1) [reply .github/config/linux.json]: This was discussed in the unit tests tech talk last week—the infra team currently just globs *tests binaries and runs those, so won't pick up nplb. You should reach out to the infra team (probably @oxve) to ask about how best to go about this or if they have something in the works

Comment (y4vor) [reply .github/config/linux.json]: Figured out where to add nplb in the .github files.

File: .github/actions/build/action.yaml [hunk 7]
```diff
@@ -65,7 +65,7 @@ runs:
         run: |
           set -x
           cd src
-          tar cvf - out/${{ matrix.platform }}_${{ matrix.config }}/*tests out/${{ matrix.platform }}_${{ matrix.config }}/*.so out/${{ matrix.platform }}_${{ matrix.config }}/*.so.* out/${{ matrix.platform }}_${{ matrix.config }}/starboard/*.so.* | xz -T0 -1 -z - > test_artifacts.tar.xz
+          tar cvf - out/${{ matrix.platform }}_${{ matrix.config }}/*tests out/${{ matrix.platform }}_${{ matrix.config }}/nplb out/${{ matrix.platform }}_${{ matrix.config }}/*.so out/${{ matrix.platform }}_${{ matrix.config }}/*.so.* out/${{ matrix.platform }}_${{ matrix.config }}/starboard/*.so.* out/${{ matrix.platform }}_${{ matrix.config }}/scontent/test | xz -T0 -1 -z - > test_artifacts.tar.xz
```
Comment (andrewsavage1) [inline .github/actions/build/action.yaml:68]: I think this is fine to get nplb in but @oxve the infra team should figure out how to maintain this long-term (or let us know if you have an immediate solution)

**Review by andrewsavage1 — APPROVED** [review]

File: starboard/common/test_main.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (andrewsavage1) [reply starboard/common/test_main.cc:39]: Is it appropriate for android to use this as an entry point or should the setup for it be different?

**Review by niranjanyardi — APPROVED** [review]

**Review by TyHolc — APPROVED** [review]

File: starboard/common/test_main.cc [hunk 8]
```diff
@@ -27,12 +27,19 @@ int InitAndRunAllTests(int argc, char** argv) {
...
 }  // namespace
 
-// When we are building Evergreen we need to export SbEventHandle so that the
-// ELF loader can find and invoke it.
+#if BUILDFLAG(IS_STARBOARD)
+// For the Starboard OS define SbEventHandle as the entry point
 SB_EXPORT STARBOARD_WRAP_SIMPLE_MAIN(InitAndRunAllTests);
+
+#if !SB_IS(EVERGREEN)
+// Define main for non-Evergreen Starboard OS.
```
Comment (yell0wd0g) [inline starboard/common/test_main.cc:35]: micro-nit: `main()` since it refers to a function.

File: starboard/common/test_main.cc [hunk 9]
```diff
@@ -27,12 +27,19 @@ int InitAndRunAllTests(int argc, char** argv) {
...
+int main(int argc, char** argv) {
+  SbRunStarboardMain(argc, argv, SbEventHandle);
+}
+#endif  // !SB_IS(EVERGREEN)
 #else
+// If the OS is not Starbaord use the regular main e.g. ATV
```
Comment (yell0wd0g) [inline starboard/common/test_main.cc:41]: nit: s/Starbaord/Starboard/ (typo). Also period at the end is missing.

File: starboard/common/test_main.cc [hunk 10]
```diff
@@ -27,12 +27,19 @@ int InitAndRunAllTests(int argc, char** argv) {
...
 int main(int argc, char** argv) {
-  ::testing::InitGoogleTest(&argc, argv);
-  return RUN_ALL_TESTS();
+  InitAndRunAllTests(argc, argv);
```
Comment (yell0wd0g) [inline starboard/common/test_main.cc:43]: nit: `return InitAndRunAllTests(...);` (missing `return`). Same in l.37.

**Review by yell0wd0g — APPROVED** [review]

File: starboard/common/test_main.cc (hunk 9 above)
Comment (y4vor) [reply starboard/common/test_main.cc:41]: deon

File: starboard/common/test_main.cc (hunk 10 above)
Comment (y4vor) [reply starboard/common/test_main.cc:43]: Done

File: starboard/common/test_main.cc (hunk 8 above)
Comment (y4vor) [reply starboard/common/test_main.cc]: Done

**Pull request merged**